### PR TITLE
[ONNX] Trace model if quantization is detected

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #76055
* #76002
* __->__ #75921
* #75920

Previously pre-tracing model is required for exporting quantized model.
e.g. calling `traced_m = torch.jit.trace(model, inputs)` and export `traced_m`.
The reason was quantized weights are stored in a unique `PackedParam` structure,
and they need to be handled by tracing to be exportable.
This PR enables export api to call tracing underneath if it detects quantization
in the model.

### CODE REVIEW

File: torch/onnx/utils.py [hunk 1]
```diff
@@ -495,6 +495,19 @@ def unpack_quantized_tensor(value):
...
 
 
+def _pre_trace_quant_model(model, args):
+    r"""Returns `torch.jit.trace(model, args)` if model is quantized. Otherwise do nothing and return
+    original model.
+
+    This is because the quantized weights are stored in special `PackedParams` structure. They need to
+    be handled by `torch.jit.trace` to be exportable to ONNX.
+    """
+    if (any([hasattr(m, "_packed_params") for m in getattr(model, "modules", lambda: [])()]) or
```
Comment (garymm) [inline torch/onnx/utils.py:505]: I don't think you need a list inside `any`:
```suggestion
    if (any(hasattr(m, "_packed_params") for m in getattr(model, "modules", lambda: [])()) or
```

File: torch/onnx/utils.py [hunk 2]
```diff
@@ -495,6 +495,19 @@ def unpack_quantized_tensor(value):
...
+            any([getattr(arg, "is_quantized", False) for arg in args])):
```
Comment (garymm) [inline torch/onnx/utils.py:506]:
```suggestion
            any(getattr(arg, "is_quantized", False) for arg in args)):
```

File: torch/onnx/utils.py [hunk 3]
```diff
@@ -573,7 +587,11 @@ def _model_to_graph(model, args, verbose=False,
             output_wrapped = torch_out  # type: ignore[assignment]
 
         output_tensors, out_desc = torch._C._jit_flatten(tuple(output_wrapped))
-        torch._C._jit_pass_onnx_assign_output_shape(graph, output_tensors, out_desc, _onnx_shape_inference)
+        # NOTE: assign_output_shape pass is not compatible with quantized outputs.
+        #       Quantized outputs are flattened to 3 values in ONNX, while packed as
+        #       single value in PyTorch.
+        if not any([getattr(out, "is_quantized", False) for out in output_tensors]):
```
Comment (garymm) [inline torch/onnx/utils.py:593]: ditto on list inside any

File: torch/onnx/utils.py [hunk 4]
```diff
@@ -573,7 +587,11 @@ def _model_to_graph(model, args, verbose=False,
             output_wrapped = torch_out  # type: ignore[assignment]
 
         output_tensors, out_desc = torch._C._jit_flatten(tuple(output_wrapped))
-        torch._C._jit_pass_onnx_assign_output_shape(graph, output_tensors, out_desc, _onnx_shape_inference)
+        # NOTE: assign_output_shape pass is not compatible with quantized outputs.
```
Comment (garymm) [inline torch/onnx/utils.py:590]: Nit: Remove "NOTE:". The comment will be shorter and have the same informational content without it.

File: torch/onnx/utils.py [hunk 5]
```diff
@@ -495,6 +495,19 @@ def unpack_quantized_tensor(value):
         return (value,)
 
 
+def _pre_trace_quant_model(model, args):
+    r"""Returns `torch.jit.trace(model, args)` if model is quantized. Otherwise do nothing and return
+    original model.
+
+    This is because the quantized weights are stored in special `PackedParams` structure. They need to
+    be handled by `torch.jit.trace` to be exportable to ONNX.
```
Comment (garymm) [inline torch/onnx/utils.py:503]: Please add a link to an issue that we're using to track getting rid of the need for this.

I guess https://github.com/pytorch/pytorch/issues/47887 ?

**Review by garymm — APPROVED** [review]

File: torch/onnx/utils.py (hunk 5 above)
Comment (BowenBao) [reply torch/onnx/utils.py:503]: #47887 is a different issue. The closest tracking one is #75761. I will link it here and leave it open then.

**Pull request closed without merging**